event_sourcery_sqlalchemy/subscription.py

```diff
@@ -30,4 +54,90 @@ def subscribe_to_events(
         timelimit: timedelta,
         events: list[str],
     ) -> Iterator[list[RecordedRaw]]:
-        raise NotImplementedError
+        return GapDetectingIterator(
+            session=self._session,
+            gap_retry_interval=self.GAP_RETRY_INTERVAL,
+            start_from=start_from,
+            batch_size=batch_size,
+            timelimit=timelimit,
+            filtering=models.Event.name.in_(events),
+        )
+
+
+@dataclass
+class Cursor:
+    position: Position
+
+    def advance(self, batch: list[models.Event]) -> None:
+        if len(batch) > 0:
+            self.position = batch[-1].id
+
+
+class GapDetectingIterator(Iterator[list[RecordedRaw]]):
+    def __init__(
+        self,
+        session: Session,
+        gap_retry_interval: timedelta,
+        start_from: Position,
+        batch_size: int,
+        timelimit: timedelta,
+        filtering: Any | None = None,  # TODO: _ColumnExpressionArgument[bool]
+    ) -> None:
+        self._session = session
+        self._gap_retry_interval = gap_retry_interval
+        self._cursor = Cursor(position=start_from)
+        self._batch_size = batch_size
+        self._timelimit = timelimit
+        self._filtering = filtering
+
+    def __next__(self) -> list[RecordedRaw]:  # TODO: duplicated from django impl
+        start = time.monotonic()
+        while True:
+            batch = self._get_batch()
+            if self._is_continuous(batch) and len(batch) == self._batch_size:
+                self._cursor.advance(batch)
+                return self._batch_to_recorded_raw(batch)
+            elif time.monotonic() - start > self._timelimit.total_seconds():
+                self._cursor.advance(batch)
+                return self._batch_to_recorded_raw(batch)
+            else:
+                time.sleep(self._gap_retry_interval.total_seconds())
+
+    def _get_batch(self) -> list[models.Event]:
+        stmt = (
+            select(models.Event)
+            .join(models.Event.stream)
+            .where(models.Event.id > self._cursor.position)
+            .order_by(models.Event.id)
+            .limit(self._batch_size)  # TODO: is ORM stmt.limit() ok?
```

Comment:
WDYM? I don't like that statement is created in the GapDetection part, but I don't see the problem with limit here.

Reply:
TODO comment is removed